pkg/clio/errors.go

```diff
@@ -0,0 +1,76 @@
+package clio
+
+type Printer interface {
+	Print()
+}
+
+type PrintCLIErrorer interface {
+	PrintCLIError()
+}
+type CLIError struct {
+	// Err is a string to avoid puntuation linting
+	Err string
+	// setting ExcludeDefaultError to true will cause the PrintCLIError method to only print the contents of Messages
+	// by default PrintCLIError first prints the contents of Err, followed by the contents of Messages
+	ExcludeDefaultError bool
+	// Messages are items which implement the clio.Printer interface.
+	// when PrintCLIError is called, each of the messages Print() method is called in order of appearence in the slice
+	Messages []Printer
+}
+
+type LogMsg struct {
+	Msg string
+}
+
+func (m *LogMsg) Print() {
+	Log(m.Msg)
+}
+
+type InfoMsg struct {
+	Msg string
+}
+
+func (m *InfoMsg) Print() {
+	Info(m.Msg)
+}
+
+type WarnMsg struct {
+	Msg string
+}
+
+func (m *WarnMsg) Print() {
+	Warn(m.Msg)
+}
+
+type DebugMsg struct {
+	Msg string
+}
+
+func (m *DebugMsg) Print() {
+	Debug(m.Msg)
+}
+
+// Error implements the error interface. It uses the default message of the
+// wrapped error.
+func (e *CLIError) Error() string {
+	return e.Err
+}
+
+// PrintCLIError prints the error message and then any messages in order from the slice
+// The indended use is to surface errors with useful messages then os.Exit without having to place os.Exit withing methods other than the cli main function
+//
+// err := CLIError{Err: errors.New("new error"), Messages: []Printer{&LogMsg{Msg:"hello world"}}}
```

Comment:
This might be a nitpick, but do we need to have the nested Msg struct on LogMsg. Would become a bit more readable if it was just a string. 
Unless you had plans to expand the LogMsg struct to include other fields.

Reply:
> This might be a nitpick, but do we need to have the nested Msg struct on LogMsg. Would become a bit more readable if it was just a string. Unless you had plans to expand the LogMsg struct to include other fields.



> This might be a nitpick, but do we need to have the nested Msg struct on LogMsg. Would become a bit more readable if it was just a string. Unless you had plans to expand the LogMsg struct to include other fields.

Good call Jack,



I had a go and looks like this will work just fine


``` go
&clio.CLIError{
Err: "Failed to load AWS credentials.",
Messages: []clio.Printer{
	clio.DebugMsg(fmt.Sprintf("Encountered error while loading default aws config: %s", err)),
},

```

``` go
type LogMsg string

func (m LogMsg) Print() {
	Log(string(m))
}

type InfoMsg string

func (m InfoMsg) Print() {
	Info(string(m))
}

type WarnMsg string

func (m WarnMsg) Print() {
	Warn(string(m))
}

type DebugMsg string

func (m DebugMsg) Print() {
	Debug(string(m))
}
```